### DISCUSSION TITLE
TPM vs NOPM

### DISCUSSION THREAD

**Discussion by az-2022** [TPC-Council/HammerDB#197] [Q&A] [answered]
Hi.  I'm a noob with a benchmarking level 101 question. I want to use HammerDB in a SQL Server environment.  What is meant by TPM metric is database specific vs NOPM metric is database independent?  Can you use a simple example to illustrate that concept?  Thanks.

**Comment by sm-shaw** [comment, marked as answer]
There is a blog post here that can give more details https://www.hammerdb.com/blog/uncategorized/why-both-tpm-and-nopm-performance-metrics/ but this is a good question so as a basic overview here as well NOPM is the primary metric. This is the equivalent of the official tpmC metric for TPC-C. As HammerDB is a derived  workload it cannot use tpmC and therefore uses the NOPM metric instead which records the new orders processed per minute (which is the same as tpmC).  Therefore NOPM is the metric that should be used in particular when comparing HammerDB results between different databases.  To be more specific NOPM can bw queried from the schema using` select sum(d_next_o_id) from district` however we do not want to be sampling this metric during a test as it could cause locking that impacts the workload itself, therefore HammerDB only queries this at the start and end. (The connect pool workload is an exception and also reports client side metrics as well for all of the stored procedures). 
TPM on the other hand is complementary and more of a database engineers metric to help you drill down into your database performance for a specific database engine. For SQL Server this is running `select cntr_value from sys.dm_os_performance_counters where counter_name = 'Batch Requests/sec'` and if you run SSMS and look at Batch/Requests per sec this is exactly the same metric that HammerDB uses. This therefore helps you relate the HammerDB workload to the database specific metrics but you cannot compare SQL Server TPM to (for Example) Db2 TPM as they uses different metrics internally. Also TPM is an in-memory metric so you can safely sample this to monitor your workload for all databases without running anything against the schema being tested. 
Therefore in summary NOPM and TPM are complementary, NOPM is the key metric and tpmC equivalent used for comparing results, TPM is the metric that helps you drill down into performance and compare your transaction rate to other database metrics for a specific database engine.

**Reply by az-2022** [reply]
Thank your for the concise reply and for the blog article.

**Reply by theBeginner86** [reply]
Thank you! This is helpful.

**Reply by sm-shaw** [reply]
Also see updated blog post here : https://www.hammerdb.com/blog/uncategorized/how-to-understand-tpc-c-tpmc-and-tproc-c-nopm-and-what-is-good-performance/
